lib/web_ui/lib/src/engine/canvaskit/surface_factory.dart

```diff
@@ -81,6 +81,42 @@ class SurfaceFactory {
     }
   }
 
+  /// Releases all surfaces so they can be reused in the next frame.
+  ///
+  /// If a released surface is in the DOM, it is not removed. This allows the
+  /// engine to release the surfaces at the end of the frame so they are ready
+  /// to be used in the next frame, but still used for painting in the current
+  /// frame.
+  void releaseSurfaces() {
+    _cache.addAll(_liveSurfaces);
+    _liveSurfaces.clear();
+  }
+
+  /// Removes all surfaces except the base surface from the DOM.
+  ///
+  /// This is called at the beginning of the frame to prepare for painting into
+  /// the new surfaces.
+  void removeSurfacesFromDom() {
+    for (final Surface surface in _cache) {
+      if (_isInDom(surface)) {
+        _removeFromDom(surface);
+      }
+    }
+    if (_isInDom(backupSurface)) {
+      _removeFromDom(backupSurface);
+    }
+  }
+
+  // Returns [true] if this [surface] is in the DOM.
+  static bool _isInDom(Surface surface) {
+    return surface.htmlElement.parent != null;
```

Comment:
Not sure if it's any better but there's also the `isConnected`: https://developer.mozilla.org/en-US/docs/Web/API/Node/isConnected

I'm also getting a feeling that the fact that we're reading back from the DOM indicates a design issue somewhere. I feel like we should know whether the surface is in use or not without the readback 🤔 Can't think of any specific harm in this; just a feeling.

Reply:
I only included it so that I could quickly remove all overlays from the DOM before recreating the scene. A quick check shows that calling `remove` on an Element that isn't in the DOM is fine, so I'll just get rid of this check.